karapace/kafka_admin.py

```diff
@@ -0,0 +1,290 @@
+
+    def get_topic_config(
+        self,
+        name: str,
+        config_name_filter: Container[str] | None = None,
+        config_source_filter: Container[ConfigSource] | None = None,
+    ) -> dict[str, str]:
+        """Fetches, filters and returns topic configuration.
+
+        The two filters, `config_name_filter` and `config_source_filter` work together
+        so if a config entry matches either of them, it'll be returned.
+        If a filter is not provided (ie. is `None`), it'll act as if matching all
+        config entries.
+        """
+        LOG.info(
+            "Fetching config for topic '%s' with name filter %s and source filter %s",
+            name,
+            config_name_filter,
+            config_source_filter,
+        )
+        futmap: dict[ConfigResource, Future] = self.describe_configs([ConfigResource(ResourceType.TOPIC, name)])
+        try:
+            topic_configs = single_futmap_result(futmap)
+        except KafkaException as exc:
+            raise_from_kafkaexception(exc)
+
+        config: dict[str, str] = {}
+        for config_name, config_entry in topic_configs.items():
+            matches_name_filter: bool = config_name_filter is None or config_name in config_name_filter
+            matches_source_filter: bool = (
+                config_source_filter is None or ConfigSource(config_entry.source) in config_source_filter
+            )
+
+            if matches_name_filter or matches_source_filter:
+                config[config_name] = config_entry.value
+
+        return config
+
+    def get_offsets(self, topic: str, partition_id: int) -> dict[str, int]:
+        """Returns the beginning and end offsets for a topic partition.
+
+        Making two separate requests for beginning and end offsets, due to the
+        `AdminClient.list_offsets` behaviour: it expects a dictionary of topic
+        partitions as keys, thus unable to fetch different values in one request
+        for the same topic and partition.
+        """
+        try:
+            LOG.info("Fetching latest offset for topic '%s' partition %s", topic, partition_id)
+            latest_offset_futmap: dict[TopicPartition, Future] = self.list_offsets(
+                {
+                    TopicPartition(topic, partition_id): OffsetSpec.latest(),
+                }
+            )
+            endoffset = single_futmap_result(latest_offset_futmap)
+
+            LOG.info("Fetching earliest offset for topic '%s' partition %s", topic, partition_id)
+            earliest_offset_futmap: dict[TopicPartition, Future] = self.list_offsets(
+                {
+                    TopicPartition(topic, partition_id): OffsetSpec.earliest(),
+                }
+            )
+            startoffset = single_futmap_result(earliest_offset_futmap)
+        except KafkaException as exc:
+            code = exc.args[0].code()
+            if code < 0:
+                raise UnknownTopicOrPartitionError() from exc
```

Comment:
issue: This looks off to me, the two codes that are `< 0` are `NONE` (no error) and `UNKNOWN_SERVER_ERROR`: https://kafka.apache.org/protocol.html#protocol_error_codes

Reply:
Indeed, `-1` is the unknown error.
However, `confluent_kafka` has its own error codes, eg. in some cases this request would throw a `_NOENT` [KafkaError](https://docs.confluent.io/platform/current/clients/confluent-kafka-python/html/index.html#kafkaerror) with the code `-156` and `"Failed to query partition leaders: No leaders found"` message.

I'll change this to explicitly check for that error code instead.